### PR TITLE
Perturbed optimizers notebook and narrative documentation.

### PR DESCRIPTION
**Pull request by q-berthet**

### CODE REVIEW

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/google/jaxopt/blob/main/docs/notebooks/perturbed_optimizers/perturbed_optimizers.ipynb)
+
++++
+
+We propose in this work a universal method to transform any optimizer $y^*$ in a differentiable approximation $y_\varepsilon^*$, using pertutbations following the method of [Berthet et al. (2020)](https://arxiv.org/abs/2002.08676). We provide a JAX implementation, illustrated here on some examples.
+
+Concretely, for an optimizer function $y^*$ defined by
+
+$$y^*(\theta) = \mathop{\mathrm{arg\,max}}_{y\in \mathcal{C}} \langle y, \theta \rangle$$
```
Comment (mblondel) [inline docs/notebooks/perturbed_optimizers/perturbed_optimizers.md:37]:
```suggestion
$$y^*(\theta) = \mathop{\mathrm{arg\,max}}_{y\in \mathcal{C}} \langle y, \theta \rangle,$$
```

**Review by mblondel — COMMENTED** [review]
Good job @q-berthet and @fabianp. Some comments below.

File: docs/perturbations.rst [hunk 2]
```diff
@@ -0,0 +1,82 @@
+Perturbed optimization
+======================
+
+The perturbed optimization module allows to transform a non-smooth function such as a max or arg-max into a differentiable function using random perturbations. This is useful for optimization algorithms that require differentiability, such as gradient descent.
+
```
[the remark this answers is not in the document]
Comment (fabianp) [reply docs/perturbations.rst:5]: good idea!

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+# Perturbed optimizers
+
++++
+
+[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/google/jaxopt/blob/main/docs/notebooks/perturbed_optimizers/perturbed_optimizers.ipynb)
+
++++
+
+We propose in this work a universal method to transform any optimizer $y^*$ in a differentiable approximation $y_\varepsilon^*$, using pertutbations following the method of [Berthet et al. (2020)](https://arxiv.org/abs/2002.08676). We provide a JAX implementation, illustrated here on some examples.
```
[the remark this answers is not in the document]
Comment (fabianp) [reply docs/notebooks/perturbed_optimizers/perturbed_optimizers.md:33]: I would also change "We provide a JAX implementation, illustrated here on some examples." by "JAXopt provides an implementation that we illustrate here on some examples"

File: docs/perturbations.rst [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+Argmax perturbations
+--------------------
+
+Consider an arg-max function of the form:
+
+.. math::
+
+    y^\star(\theta) = \mathop{\mathrm{arg\,max}}_{y \in \mathcal{C}} \langle y, \theta\rangle\,,
```
Comment (fabianp) [inline docs/perturbations.rst:45]: please replace y^\star by y^* for consistency with the notebook and the paper

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+from jaxopt import perturbations
+```
+
++++ {"id": "EmYn_jNUFfw2"}
+
+# Argmax one-hot
+
++++
+
+We start from an original optimizer, an `argmax_one_hot` function, corresponding to $y^*$ for $\mathcal{C}$ being the unit simplex. We run it on an example input `values`
```
Comment (fabianp) [inline docs/notebooks/perturbed_optimizers/perturbed_optimizers.md:67]: What does "original" mean in this context? It is not clear to me

Comment (fabianp) [reply docs/notebooks/perturbed_optimizers/perturbed_optimizers.md:67]: I think it would also be good to explain with words what the `argmax_one_hot` does. Something along the lines of:

"We consider the `argmax_one_hot` function, which transforms a real-valued vector into a binary vector with one in the coefficient with largest magnitude and zero elsewhere"

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+print(one_hot_vec)
+```
+
++++ {"id": "6rbNt-6zGb-J"}
+
+## One-hot with pertubations
+
++++
+
+Our implementation allows to modify this function by creating a perturbed optimizer `pert_one_hot`, corresponding to $y_\varepsilon^*$ using Gumbel noise for $Z$.
```
Comment (fabianp) [inline docs/notebooks/perturbed_optimizers/perturbed_optimizers.md:99]:
```suggestion
Our implementation transforms the `argmax_one_hot` function into perturbed one that we call `pert_one_hot`. In this case we use Gumbel noise for the perturbation.
```

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+num_samples = 100000
+sigma = 0.5
+rng = jax.random.PRNGKey(1)
+pert_one_hot = perturbations.make_perturbed_argmax(argmax_fun=argmax_one_hot,
+                                         num_samples=num_samples,
+                                         sigma=sigma,
+                                         noise=perturbations.Gumbel())
+```
+
+In this particular case, it is equal to the usual softmax, with exponential weights. This is not always true, in general there is no closed form for $y_\varepsilon^*$
```
Comment (fabianp) [inline docs/notebooks/perturbed_optimizers/perturbed_optimizers.md:113]:
```suggestion
In this particular case, the perturbed argmax is equivalent to the [softmax function](https://en.wikipedia.org/wiki/Softmax_function). This is not always true, in general there is no closed form for $y_\varepsilon^*$
```

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+## One-hot with pertubations
+
++++
+
+Our implementation allows to modify this function by creating a perturbed optimizer `pert_one_hot`, corresponding to $y_\varepsilon^*$ using Gumbel noise for $Z$.
+
+```{code-cell} ipython3
+:id: 7hQz6zuPwkpZ
+
+num_samples = 100000
```
Comment (fabianp) [inline docs/notebooks/perturbed_optimizers/perturbed_optimizers.md:104]:
```suggestion
NUM_SAMPLES = 100_000
```
it's a convention to write global constants in ALL_CAPS

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+sigma = 0.5
```
Comment (fabianp) [inline docs/notebooks/perturbed_optimizers/perturbed_optimizers.md:105]:
```suggestion
SIGMA = 0.5
```

(and adapt the code after to use these new variable names)

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+
++++ {"id": "2U7rhtEAGpMV"}
+
+## Gradients for one-hot with perturbations
+
++++
+
+The perturbed optimizer $y_\varepsilon^*$ is differentiable, and its gradient can be computed with stochastic estimation automatically, using `jax.grad`.
+
+We create a scalar loss `loss_simplex` of the perturbed optimizer $y^*_\varepsilon$
```
Comment (fabianp) [inline docs/notebooks/perturbed_optimizers/perturbed_optimizers.md:144]: perhaps worth saying that this is a squared loss between y_true and y_eps ? This was not clear to me from the text + math

**Review by fabianp — CHANGES_REQUESTED** [review]

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (q-berthet) [reply docs/notebooks/perturbed_optimizers/perturbed_optimizers.md]: done

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md (hunk 1 above)
Comment (q-berthet) [reply docs/notebooks/perturbed_optimizers/perturbed_optimizers.md]: done

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md (hunk 5 above)
Comment (q-berthet) [reply docs/notebooks/perturbed_optimizers/perturbed_optimizers.md:67]: done

File: docs/perturbations.rst (hunk 4 above)
Comment (q-berthet) [reply docs/perturbations.rst:45]: done

File: docs/perturbations.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (q-berthet) [reply docs/perturbations.rst:5]: done

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md (hunk 1 above)
Comment (q-berthet) [reply docs/notebooks/perturbed_optimizers/perturbed_optimizers.md]: done

File: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (q-berthet) [reply docs/notebooks/perturbed_optimizers/perturbed_optimizers.md]: done

**Comment by q-berthet** [comment]
Thanks! All done.

**Review by fabianp — APPROVED** [review]

**Comment by fabianp** [comment]
congrats @q-berthet !
